hax/hax/handler.py

```diff
@@ -61,23 +61,70 @@ def stop(self) -> None:
         self.is_stopped = True
 
     @repeat_if_fails(wait_seconds=1)
-    def _update_process_status(self, event: ConfHaProcess) -> None:
+    def _update_process_status(self, q: Queue, event: ConfHaProcess) -> None:
         # If a consul-related exception appears, it will
         # be processed by repeat_if_fails.
         #
         # This thread will become blocked until that
         # intermittent error gets resolved.
         self.consul.update_process_status(event)
+        svc_status = m0HaProcessEvent.event_to_svchealth(event.chp_event)
+        if event.chp_type == m0HaProcessType.M0_CONF_HA_PROCESS_M0D:
+            # Broadcast the received motr process status to other motr
+            # processes in the cluster.
+            q.put(BroadcastHAStates(states=[HAState(fid=event.fid,
+                                    status=svc_status)],
+                  reply_to=None))
 
-    def update_process_failure(self, ha_states: List[HAState]) -> None:
+    def update_process_failure(self, q: Queue,
+                               ha_states: List[HAState]) -> List[HAState]:
+        new_ha_states: List[HAState] = []
         for state in ha_states:
-            if state.status == ServiceHealth.FAILED:
-                m0status = m0HaProcessEvent.M0_CONF_HA_PROCESS_STOPPED
-                pevent = ConfHaProcess(chp_event=m0status,
-                                       chp_type=3,
-                                       chp_pid=0,
-                                       fid=state.fid)
-                self._update_process_status(pevent)
+            if self.consul.fid2objT(str(state.fid)) == 'PROCESS':
+                svc_status = state.status
+                if state.status == ServiceHealth.FAILED:
+                    # check one more time before broadcasting failure.
+                    node = self.consul.get_process_node(state.fid)
+                    proc_status = self.consul.get_process_status(state.fid)
+                    svc_status = self.consul.get_service_health(node,
+                                                                state.fid.key)
+                    LOG.info('node: %s proc: %s proc_state: %s health: %s',
+                             node, state.fid, proc_status, svc_status)
+                    if svc_status == ServiceHealth.FAILED:
+                        m0status = m0HaProcessEvent.M0_CONF_HA_PROCESS_STOPPED
+                        if proc_status != 'M0_CONF_HA_PROCESS_STOPPED':
+                            pevent = ConfHaProcess(chp_event=m0status,
+                                                   chp_type=3,
```

Comment:
what `chp_type=3` stands for?

Reply:
Its `m0HaProcessType.M0_CONF_HA_PROCESS_M0D`, updated.